File: 3827-implement-router/3827-implement-router.cpp
class Router {
public:
    Router(int memoryLimit) : memoryLimit(memoryLimit) {
        
    }
    
    bool addPacket(int source, int destination, int timestamp) {
        string packet = to_string(source) + "*" + to_string(destination) + "*" + to_string(timestamp);
        if(packetHash.find(packet) != packetHash.end())
        {
            return false;
        }
        if(packets.size() == memoryLimit)
        {
            int d = packets.front();
            packets.pop();
            int s = dests[d].front().second;
            int ts = dests[d].front().first;
            dests[d].pop_front();
            string p = to_string(s) + "*" + to_string(d) + "*" + to_string(ts);
            packetHash.erase(p);
        }
        dests[destination].push_back({timestamp, source});
        packets.push(destination);
        packetHash.insert(packet);
        return true;
    }
    
    vector<int> forwardPacket() {
        if(packets.empty())
        {
            return {};
        }
        int destination = packets.front();
        packets.pop();
        int source = dests[destination].front().second;
        int timestamp = dests[destination].front().first;
        dests[destination].pop_front();
        string packet = to_string(source) + "*" + to_string(destination) + "*" + to_string(timestamp);
        packetHash.erase(packet);
        return {source, destination, timestamp};
    }
    
    int getCount(int destination, int startTime, int endTime) {
        int l1 = 0;
        int r1 = dests[destination].size();
        while(l1 < r1)
        {
            int m = (l1 + r1) / 2;
            if(dests[destination][m].first < startTime)
            {
                l1 = m + 1;
            }
            else
            {
                r1 = m;
            }
        }
        if(r1 == dests[destination].size() || dests[destination][r1].first > endTime)
        {
            return 0;
        }
        int l2 = 0;
        int r2 = dests[destination].size();
        while(l2 < r2)
        {
            int m = (l2 + r2 + 1) / 2;
            if(m < dests[destination].size() && dests[destination][m].first <= endTime)
            {
                l2 = m;
            }
            else
            {
                r2 = m - 1;
            }
        }
        return r2 - r1 + 1;
    }
private:
    int memoryLimit;
    unordered_set<string> packetHash;
    queue<int> packets;
    unordered_map<int, deque<pair<int, int>>> dests;
};

/**
 * Your Router object will be instantiated and called as such:
 * Router* obj = new Router(memoryLimit);
 * bool param_1 = obj->addPacket(source,destination,timestamp);
 * vector<int> param_2 = obj->forwardPacket();
 * int param_3 = obj->getCount(destination,startTime,endTime);
 */

File: 3827-implement-router/3827-implement-router.py
class Router:

    def __init__(self, memoryLimit: int):
        self.memoryLimit = memoryLimit
        self.packetHash = set()
        self.packets = deque()
        self.dests = {} 

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:
        packet = f"{source}*{destination}*{timestamp}"
        if packet in self.packetHash:
            return False
        if len(self.packets) == self.memoryLimit:
            d = self.packets.popleft()
            ts, s = self.dests[d].popleft()          
            p = f"{s}*{d}*{ts}"
            self.packetHash.remove(p)
        if destination not in self.dests:
            self.dests[destination] = deque()
        self.dests[destination].append((timestamp, source))
        self.packets.append(destination)
        self.packetHash.add(packet)
        return True
        
    def forwardPacket(self) -> List[int]:
        if len(self.packets) == 0:
            return []
        destination = self.packets.popleft()
        timestamp, source = self.dests[destination].popleft()
        packet = f"{source}*{destination}*{timestamp}"
        self.packetHash.remove(packet)
        return [source, destination, timestamp]

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:
        if destination not in self.dests:
            return 0
        l1 = 0
        r1 = len(self.dests[destination])
        while l1 < r1:
            m = (l1 + r1) // 2
            if self.dests[destination][m][0] < startTime:
                l1 = m + 1
            else:
                r1 = m
        if r1 == len(self.dests[destination]) or self.dests[destination][r1][0] > endTime:
            return 0
        l2 = 0
        r2 = len(self.dests[destination])
        while l2 < r2:
            m = (l2 + r2 + 1) // 2
            if m < len(self.dests[destination]) and self.dests[destination][m][0] <= endTime:
                l2 = m
            else:
                r2 = m - 1
        return r2 - r1 + 1
        


# Your Router object will be instantiated and called as such:
# obj = Router(memoryLimit)
# param_1 = obj.addPacket(source,destination,timestamp)
# param_2 = obj.forwardPacket()
# param_3 = obj.getCount(destination,startTime,endTime)